posts/2022/05/the-evolution-of-the-scipy-developer-cli.md

```diff
@@ -0,0 +1,170 @@
+<!--
+.. title: The evolution of the SciPy developer CLI
+.. slug: the-evolution-of-the-scipy-developer-cli
+.. date: 2022-05-03 8:00:00 UTC+05:30
+.. author: Sayantika Banik
+.. tags: CLI, SciPy, doit, Typer, Rich, Click
+.. category:
+.. link:
+.. description:
+.. type: text
+.. previewimage:
+-->
+
+#### 🤔 What is a command-line interface (CLI)?
+
+Imagine a situation where there is a massive system with various tools and functionality. Every functionality requires a special command or an input from the user, CLI is designed to tackle such situations. Like a catalog it lists all the options present, helping the user to navigate.
+
+<p align="center">
+  <img alt="CLI example" src="https://user-images.githubusercontent.com/17350312/166633508-a2795c44-30bc-4a5b-8043-65beab71d31f.png" />
+</p>
+
+Now that we understand what a `CLI` is, how about we dive into the world of `SciPy`?
+
+<!-- TEASER_END -->
+
+#### 🤓 SciPy CLI journey
+
+An open-source project undergoes multiple iterations, and each iteration adds layers of tools and functionalities to incorporate. Right from the initial build, test, benchmarking, release notes, and beyond. Contribution guidelines start spanning across multiple pages, exponentially increasing efforts to maintain.
+
+The two homegrown CLI for SciPy are `runtests.py`(for distutils-based builds) and `dev.py`(for Meson-based builds) built using python tooling `argparse`.
+Both of these toolings have multiple conditional blocks to achieve the desired outcome. Not only it is difficult to maintain, overall readability decreases drastically. The documentation runs into an infinite loop of updates, requiring additional efforts from maintainers. Another issue that remains is the lack of task grouping. A group of tasks linking to a single objective helps reduce confusion, super helpful for new contributors (like me :D).
+
+Hence the idea of a developer command-line interface (CLI) was born. Easing the development experience with an intuitive and informative CLI. Removing dependency on legacy tooling like `paver` also added great value to the overall experience. Additional details could be found under [issue-#15489](https://github.com/scipy/scipy/issues/15489)
+
+#### ✍️ Planning and objective
+Like any development activity, the goal was to experiment with tools available [doit](https://pydoit.org/) and [Typer](https://typer.tiangolo.com/) were the first ones we picked. The two components of our interests were a *task runner*, and a *command-line interface tool*. doit satisfied the requirements of a task runner along with added functionality like maintaining a task dependency graph as a `DAG`. While `Typer` is quick to get started with building CLI applications.
```

Comment:
```suggestion
Like any development activity, the plan was to experiment with available tools. [doit](https://pydoit.org/) and [Typer](https://typer.tiangolo.com/) were the first ones we picked. The two components of our interests were a *task runner*, and a *command-line interface tool*. doit satisfied the requirements of a task runner along with added functionality, like maintaining a task dependency graph as a `DAG`. While `Typer` is quick to get started with building CLI applications.
```

Reply:
I'm not sure if DAG is very familiar to people. Maybe link it to wiki?

Reply:
Sounds good. 
I found this article easy to understand - https://hazelcast.com/glossary/directed-acyclic-graph/ 
The wiki one seemed very detailed, https://en.wikipedia.org/wiki/Directed_acyclic_graph